pkg/kubelet/pod/mirror_manager.go

```diff
@@ -0,0 +1,393 @@
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	key := getKey(staticPod)
+	if _, found := m.staticPods[key]; !found {
+		glog.V(4).Infof("The static pod %q is not found", format.Pod(staticPod))
+		return
+	}
+	glog.V(2).Infof("Delete static pod %q, delete the mirror pod", format.Pod(staticPod))
+	delete(m.staticPods, key)
+	// In fact using static pod directly is also ok, but it's more intuitive making sure that
+	// only mirror pod is passing into deleteMirrorPod() and finally handleMirrorPodDeletion()
+	m.deleteMirrorPod(newMirrorPod(staticPod))
+}
+
+func (m *mirrorPodManager) AddMirrorPod(mirrorPod *api.Pod) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	// Invalid mirror pod will be recreated in periodic cleanup if needed
+	m.mirrorPods.setPod(getKey(mirrorPod), mirrorPod)
+}
+
+func (m *mirrorPodManager) DeleteMirrorPod(mirrorPod *api.Pod) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	// Removed mirror pod will be recreated in periodic cleanup if needed
+	m.mirrorPods.removePod(getKey(mirrorPod))
+}
+
+func (m *mirrorPodManager) GetMirrorPod(name, namespace string) *api.Pod {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	return m.mirrorPods.getPod(getKeyByName(name, namespace))
+}
+
+func (m *mirrorPodManager) GetMirrorPodByUID(uid types.UID) *api.Pod {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	return m.mirrorPods.getPodByUID(uid)
+}
+
+func (m *mirrorPodManager) GetMirrorOfStaticPod(staticPod *api.Pod) *api.Pod {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	return m.mirrorPods.getPod(getKey(staticPod))
+}
+
+func (m *mirrorPodManager) TranslatePodUID(uid types.UID) types.UID {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	mirrorPod := m.mirrorPods.getPodByUID(uid)
+	if mirrorPod != nil {
+		if staticPod, found := m.staticPods[getKey(mirrorPod)]; found {
+			return staticPod.UID
+		}
+	}
+	return uid
+}
+
+func IsStaticPod(pod *api.Pod) bool {
+	source, err := kubetypes.GetPodSource(pod)
+	return err == nil && source != kubetypes.ApiserverSource
+}
+
+func IsMirrorPod(pod *api.Pod) bool {
+	_, ok := pod.Annotations[kubetypes.ConfigMirrorAnnotationKey]
+	return ok
+}
+
+// Create or recreate mirror pod passed in
+func (m *mirrorPodManager) handleMirrorPodCreation(mirrorPod *api.Pod) {
+	// There is one case will cause the mirror pod manager to keep recreating mirror pod:
+	//  * Recreate a mirror pod m1, because the mirror pod now m0 is invalid.
+	//  * Receive the REMOVE message of m0 and clear the corresponding mirror pod cache.
+	//  * Can't find corresponding mirror pod in cleanup, recreate a new one m2.
+	//  * Receive the ADD and REMOVE message of m1 and clear the corresponding mirror pod cache.
+	//  * Can't find corresponding mirror pod in cleanup, recreate a new one m3
+	//  * ...
+	// So we should make sure the old mirror pod is indeed invalid so as to avoid recreation loop.
+	oldMirrorPod, err := m.apiserverClient.Core().Pods(mirrorPod.Namespace).Get(mirrorPod.Name)
+	if err == nil {
+		m.lock.RLock()
+		defer m.lock.RUnlock()
```

Comment:
This could hold the lock until the end of the function which includes multiple apiserver operations. Make it a separate function that doesn't perform remote operations while holding a lock.